File: two-pointers/1_Two_Sum.cpp
/**
 * @file    1_Two_Sum.cpp
 *
 * @author  Juan Diego Becerra
 * @date    2021-11-01
 * @brief   https://leetcode.com/problems/two-sum/
 *
 */

#include <vector>
#include <unordered_map>
using namespace std;

// Time: O(n) | Space: O(n)
vector<int> twoSum(vector<int> &nums, int target)
{
    unordered_map<int, int> seen; // value : index
    int i, j;

    for (i = 0; i < nums.size(); ++i)
    {
        auto found = seen.find(target - nums[i]);

        if (found == seen.end())
        {
            seen[nums[i]] = i;
        }
        else
        { // found != seen.end()
            j = found->second;
            break;
        }
    }
    return {i, j};
}

File: two-pointers/1_Two_Sum.py
"""1_Two_Sum.py

Author: Juan Diego Becerra
Date:   06-08-2022
Brief:  https://leetcode.com/problems/two-sum
"""


# Time: O(n) | Space: O(n)
def twoSum(nums, target: int):
    values = dict()

    for i, elem in enumerate(nums):
        comp = target - elem

        if comp in values:
            return [values[comp], i]
        values[elem] = i

    return []
